blog/_posts/2017-12-06-ml&pl-cn.md

```diff
@@ -0,0 +1,106 @@
+
+想象一个世界，增加一个新硬件的支持，或新的数据表示，不必对原有的系统做任何更改，写写高层代码就能轻松解决。当前数值计算语言[已经能轻松处理这些任务](https://arxiv.org/pdf/1604.03410.pdf)，我们期望机器学习系统能从中汲取灵感。
+
+我们知道类型系统有安全性的优点，但当前的系统却不能适应充斥着数组的代码，因为数组的维度是有意义的（例如图像中有空间，通道和批次三种维度）。维度的区分[全靠约定](https://github.com/pytorch/pytorch/issues/1220)，置换维度的繁杂代码无法保证正确性，所以说完美支持数组的类型系统还有很大的探索空间。鉴于实践者们对交互和脚本的偏好，我们预计动态类型的趋势仍将继续，[^types]我们也希望看到更多像 [CNTK 的可选动态维度](https://cntk.ai/pythondocs/sequence.html)这样的创新。
+
+[^types]: 尽管我们知道，从彻头彻尾的动态（PyTorch 和它的 ATen 后端）到异乎寻常的静态（TensorFlow 的 XLA 和 MXNet 中，所有维度在图运算之前都已知），当前的系统内部设计已全范围覆盖。
+
+机器学习工程师们对传统[软件工程问题](https://papers.nips.cc/paper/5656-hidden-technical-debt-in-machine-learning-systems.pdf)，比如生产系统的维护与扩展，的兴趣日益增加。[机器学习编程模型](https://medium.com/@karpathy/software-2-0-a64152b37c35)让组件间抽象界限的区分和接口的设计更加困难，向后兼容性可能因为重新训练一个模型就被破坏。正如普通语言要解决这些问题，机器学习语言也有希望配备相应的解决方案，虽然怎么做还是个公开的设计问题。
+
+<div style="text-align:center">
+<a href="https://xkcd.com/1838/">
+<img height="350px" src="https://imgs.xkcd.com/comics/machine_learning_2x.png"/>
+</a>
+</div>
+<div class="desc">
+  软件工程 2.0？ <i>(摘自 XKCD)</i>
+</div>
+
+任何新语言不能回避的一个不利因素是要构建新的库生态系统，因为只有专为新运行时写的代码才能从中受益。举个例子，TensorFlow 开发者们必须用图语言重写[图像处理](https://www.tensorflow.org/api_guides/python/image)和[文件输入输出](https://www.tensorflow.org/api_docs/python/tf/TextLineReader)这些功能，而不是复用 Python 生态系统中已有的库，即使对 SciPy 这类投入巨大的项目也得忍痛割爱。虽说这是前进必经的独木桥，但机器学习实践者们不应因此把自己孤立于广大的数值与高性能计算社区之外。理想的机器学习生态系统一定是理想的数值计算生态系统，反之亦然。社区间的通力合作能让所有人的努力事半功倍。
+
+我们预计进展会来自几个方面。图 IR 及像 [XLA](https://www.tensorflow.org/performance/xla/)，[ONNX](https://github.com/onnx/onnx) 和 [NNVM](https://github.com/dmlc/nnvm) 等格式已变得日益复杂并可能去传统语言设计中寻求更多的灵感。[^chris] 再加上一些表层语法，它们或许就成了完备的编程语言。TensorFlow 的 XLA 开始向专用的编译栈方向发展，现在已经有 [TVM](http://tvmlang.org/)，[DLVM](http://dlvm.org/)，[myelin](https://github.com/google/sling/tree/master/myelin)，和其它一些推进中的项目。与此同时，[PyTorch JIT](https://github.com/pytorch/pytorch/tree/master/torch/csrc/jit)，[Gluon](https://mxnet.incubator.apache.org/api/python/gluon.html) 和 [Tangent](https://github.com/google/tangent) 等正努力让 Python 本身变成一个更好的建模语言，尽管它们面对的是巨大的挑战。刚才讨论过，机器学习是一个数值计算语言的问题，我们 Julia 社区认为这个观点也为探讨语言层面的问题奠定了良好的基石。我们会继续用[Knet](https://github.com/denizyuret/Knet.jl)，[Flux](https://fluxml.github.io/)，[Cassette](https://github.com/jrevels/Cassette.jl)，[CUDAnative](https://github.com/JuliaGPU/CUDAnative.jl)，和 [DataFlow.jl](https://github.com/MikeInnes/DataFlow.jl) 等项目去扩展新的疆土。
+
+[^chris]: 一个有趣的佐证是，Google 大脑正加大力度雇用编程语言专家，比如 [Chris Lattner](https://techcrunch.com/2017/08/14/swift-creator-chris-lattner-joins-google-brain-after-tesla-autopilot-stint/)。
+
+## 结论：一个关于机器学习的推断
+
+机器学模型已是极其通用的信息处理系统，它们创建了从未有过的高层的，更复杂的抽象；通过基本组件组合，可以实现循环，递归，高阶模型，甚至[堆栈机](https://nlp.stanford.edu/blog/hybrid-tree-sequence-neural-networks-with-spinn/)和[语言解释器](https://arxiv.org/abs/1605.06640)。机器学习是一个新的编程范式，即便它满身的数值，微分，和并行标签让人觉得有些奇怪。像其它任何工程领域一样，手中的工具对将来工作的范围和质量有着不可估量的影响。
```

Comment:
> higher-level and more complex abstractions

> 高层的，更复杂的抽象

I prefer `高階的、更複雜的抽象化`

and

> 高阶模型

> higher-order models

I will choose `高維度模型`

Reply:
I prefer "level" -> "层", "order" -> "阶".
I also thought "order" should not be "維度" (dimension), or maybe its okay in traditional Chinese, I dont know

Reply:
at least in my experience, e.g. `high-level language` I always call it 高階語言.

and I use `維度` because I try to match the context about 模型.
(e.g. 我常常會說 MNIST 的 input 是 768 維，所以 高維度 常常可以聯想到 模型。也許這只是我的習慣

Reply:
高階語言 vs 高级语言 (`high-level language`) is a traditional/simplified diff. 
(here `level` is `级` in this specific usage)

Reply:
`維度` is not exact. a `high-dimensional model` is not necessarily `higher-order`. A model that relates to second-order differentiation might called high-order.

Reply:
Oh, I got your point.